src/engine/QueryExecutionTree.cpp

```diff
@@ -139,6 +139,162 @@ nlohmann::json QueryExecutionTree::writeResultAsJson(
+      b["value"] = value;
+      b["type"] = "literal";
+      ad_utility::NumericType nt = ad_utility::NumericType(entitystr.back());
+      switch (nt) {
+        case ad_utility::NumericType::FLOAT:
+          b["datatype"] = "http://www.w3.org/2001/XMLSchema#float";
+          break;
+        case ad_utility::NumericType::DOUBLE:
+          b["datatype"] = "http://www.w3.org/2001/XMLSchema#double";
+          break;
+        case ad_utility::NumericType::DECIMAL:
+          b["datatype"] = "http://www.w3.org/2001/XMLSchema#decimal";
+          break;
+        case ad_utility::NumericType::INTEGER:
+          b["datatype"] = "http://www.w3.org/2001/XMLSchema#int";
+          break;
+      }
+    } else {
+      // The string is an iri or literal
+      if (entitystr[0] == '<') {
+        b["type"] = "iri";
+        // Strip the <> surrounding the iri
+        b["value"] = entitystr.substr(1, entitystr.size() - 2);
+      } else {
+        b["type"] = "literal";
+        // Look for a language tag
+        size_t at_pos = entitystr.rfind('@');
+        if (at_pos == std::string::npos) {
+          b["value"] = entitystr.substr(1, entitystr.size() - 2);
+        } else {
+          size_t quote_pos = entitystr.rfind('"');
+          if (quote_pos != std::string::npos && quote_pos < at_pos) {
+            b["value"] = entitystr.substr(1, quote_pos - 1);
+            b["xml:lang"] = entitystr.substr(at_pos + 1);
+          } else {
+            b["value"] = entitystr.substr(1, entitystr.size() - 2);
+          }
+        }
+      }
+    }
+    return b;
+  };
+
+  for (size_t i = offset; i < upperBound; ++i) {
+    json binding;
+    for (const auto& idx : validIndices) {
+      const auto& currentId = data(i, idx.first);
+      switch (idx.second) {
+        case ResultTable::ResultType::KB: {
+          std::optional<string> entity =
+              _qec->getIndex().idToOptionalString(currentId);
+          if (entity) {
+            string entitystr = entity.value();
+            binding[selectVars[idx.first]] = entitystrToBinding(entitystr);
+          }
+          break;
+        }
+        case ResultTable::ResultType::VERBATIM: {
+          json b;
+          b["value"] = std::to_string(currentId);
+          b["type"] = "literal";
+          b["datatype"] = "http://www.w3.org/2001/XMLSchema#int";
+          binding[selectVars[idx.first]] = b;
+        } break;
+        case ResultTable::ResultType::TEXT: {
+          json b;
+          b["value"] = _qec->getIndex().getTextExcerpt(currentId);
+          b["type"] = "literal";
+          binding[selectVars[idx.first]] = b;
+        } break;
+        case ResultTable::ResultType::FLOAT: {
+          float f;
+          std::memcpy(&f, &currentId, sizeof(float));
+          std::stringstream s;
+          s << f;
+          json b;
+          b["value"] = s.str();
+          b["type"] = "literal";
+          b["datatype"] = "http://www.w3.org/2001/XMLSchema#float";
```

Comment:
Since there is at least one other position in the code, where these are used (when the value literals are parsed), we should factor these magic constants out to a common header